File: 05-determinants/03-cramers-rule/csharp/CramersRule.cs
/**
 * 克萊姆法則 (Cramer's Rule)
 *
 * 編譯：dotnet build 或 csc CramersRule.cs
 * 執行：dotnet run 或 ./CramersRule.exe
 */

using System;  // EN: Execute a statement: using System;.
using System.Text;  // EN: Execute a statement: using System.Text;.

class CramersRule  // EN: Execute line: class CramersRule.
{  // EN: Structure delimiter for a block or scope.
    static void PrintSeparator(string title)  // EN: Execute line: static void PrintSeparator(string title).
    {  // EN: Structure delimiter for a block or scope.
        Console.WriteLine();  // EN: Execute a statement: Console.WriteLine();.
        Console.WriteLine(new string('=', 60));  // EN: Execute a statement: Console.WriteLine(new string('=', 60));.
        Console.WriteLine(title);  // EN: Execute a statement: Console.WriteLine(title);.
        Console.WriteLine(new string('=', 60));  // EN: Execute a statement: Console.WriteLine(new string('=', 60));.
    }  // EN: Structure delimiter for a block or scope.

    static void PrintMatrix(string name, double[,] M)  // EN: Execute line: static void PrintMatrix(string name, double[,] M).
    {  // EN: Structure delimiter for a block or scope.
        int n = M.GetLength(0);  // EN: Execute a statement: int n = M.GetLength(0);.
        Console.WriteLine($"{name} =");  // EN: Execute a statement: Console.WriteLine($"{name} =");.
        for (int i = 0; i < n; i++)  // EN: Loop control flow: for (int i = 0; i < n; i++).
        {  // EN: Structure delimiter for a block or scope.
            var sb = new StringBuilder("  [");  // EN: Execute a statement: var sb = new StringBuilder(" [");.
            for (int j = 0; j < M.GetLength(1); j++)  // EN: Loop control flow: for (int j = 0; j < M.GetLength(1); j++).
            {  // EN: Structure delimiter for a block or scope.
                sb.Append(M[i, j].ToString("F4").PadLeft(8));  // EN: Execute a statement: sb.Append(M[i, j].ToString("F4").PadLeft(8));.
                if (j < M.GetLength(1) - 1) sb.Append(", ");  // EN: Conditional control flow: if (j < M.GetLength(1) - 1) sb.Append(", ");.
            }  // EN: Structure delimiter for a block or scope.
            sb.Append("]");  // EN: Execute a statement: sb.Append("]");.
            Console.WriteLine(sb.ToString());  // EN: Execute a statement: Console.WriteLine(sb.ToString());.
        }  // EN: Structure delimiter for a block or scope.
    }  // EN: Structure delimiter for a block or scope.

    static void PrintVector(string name, double[] v)  // EN: Execute line: static void PrintVector(string name, double[] v).
    {  // EN: Structure delimiter for a block or scope.
        var sb = new StringBuilder($"{name} = [");  // EN: Execute a statement: var sb = new StringBuilder($"{name} = [");.
        for (int i = 0; i < v.Length; i++)  // EN: Loop control flow: for (int i = 0; i < v.Length; i++).
        {  // EN: Structure delimiter for a block or scope.
            sb.Append(v[i].ToString("F4"));  // EN: Execute a statement: sb.Append(v[i].ToString("F4"));.
            if (i < v.Length - 1) sb.Append(", ");  // EN: Conditional control flow: if (i < v.Length - 1) sb.Append(", ");.
        }  // EN: Structure delimiter for a block or scope.
        sb.Append("]");  // EN: Execute a statement: sb.Append("]");.
        Console.WriteLine(sb.ToString());  // EN: Execute a statement: Console.WriteLine(sb.ToString());.
    }  // EN: Structure delimiter for a block or scope.

    // 2×2 行列式
    static double Det2x2(double[,] A)  // EN: Execute line: static double Det2x2(double[,] A).
    {  // EN: Structure delimiter for a block or scope.
        return A[0, 0] * A[1, 1] - A[0, 1] * A[1, 0];  // EN: Return from the current function: return A[0, 0] * A[1, 1] - A[0, 1] * A[1, 0];.
    }  // EN: Structure delimiter for a block or scope.

    // 3×3 行列式
    static double Det3x3(double[,] A)  // EN: Execute line: static double Det3x3(double[,] A).
    {  // EN: Structure delimiter for a block or scope.
        return A[0, 0] * (A[1, 1] * A[2, 2] - A[1, 2] * A[2, 1])  // EN: Return from the current function: return A[0, 0] * (A[1, 1] * A[2, 2] - A[1, 2] * A[2, 1]).
             - A[0, 1] * (A[1, 0] * A[2, 2] - A[1, 2] * A[2, 0])  // EN: Execute line: - A[0, 1] * (A[1, 0] * A[2, 2] - A[1, 2] * A[2, 0]).
             + A[0, 2] * (A[1, 0] * A[2, 1] - A[1, 1] * A[2, 0]);  // EN: Execute a statement: + A[0, 2] * (A[1, 0] * A[2, 1] - A[1, 1] * A[2, 0]);.
    }  // EN: Structure delimiter for a block or scope.

    static double Determinant(double[,] A)  // EN: Execute line: static double Determinant(double[,] A).
    {  // EN: Structure delimiter for a block or scope.
        int n = A.GetLength(0);  // EN: Execute a statement: int n = A.GetLength(0);.
        if (n == 2) return Det2x2(A);  // EN: Conditional control flow: if (n == 2) return Det2x2(A);.
        if (n == 3) return Det3x3(A);  // EN: Conditional control flow: if (n == 3) return Det3x3(A);.
        throw new Exception("僅支援 2×2 和 3×3 矩陣");  // EN: Execute a statement: throw new Exception("僅支援 2×2 和 3×3 矩陣");.
    }  // EN: Structure delimiter for a block or scope.

    // 替換第 j 行
    static double[,] ReplaceColumn(double[,] A, double[] b, int j)  // EN: Execute line: static double[,] ReplaceColumn(double[,] A, double[] b, int j).
    {  // EN: Structure delimiter for a block or scope.
        int n = A.GetLength(0);  // EN: Execute a statement: int n = A.GetLength(0);.
        double[,] Aj = new double[n, n];  // EN: Execute a statement: double[,] Aj = new double[n, n];.
        for (int i = 0; i < n; i++)  // EN: Loop control flow: for (int i = 0; i < n; i++).
        {  // EN: Structure delimiter for a block or scope.
            for (int k = 0; k < n; k++)  // EN: Loop control flow: for (int k = 0; k < n; k++).
            {  // EN: Structure delimiter for a block or scope.
                Aj[i, k] = (k == j) ? b[i] : A[i, k];  // EN: Execute a statement: Aj[i, k] = (k == j) ? b[i] : A[i, k];.
            }  // EN: Structure delimiter for a block or scope.
        }  // EN: Structure delimiter for a block or scope.
        return Aj;  // EN: Return from the current function: return Aj;.
    }  // EN: Structure delimiter for a block or scope.

    // 克萊姆法則
    static double[] CramersRuleSolve(double[,] A, double[] b)  // EN: Execute line: static double[] CramersRuleSolve(double[,] A, double[] b).
    {  // EN: Structure delimiter for a block or scope.
        int n = A.GetLength(0);  // EN: Execute a statement: int n = A.GetLength(0);.
        double detA = Determinant(A);  // EN: Execute a statement: double detA = Determinant(A);.

        if (Math.Abs(detA) < 1e-10)  // EN: Conditional control flow: if (Math.Abs(detA) < 1e-10).
        {  // EN: Structure delimiter for a block or scope.
            throw new Exception("矩陣奇異");  // EN: Execute a statement: throw new Exception("矩陣奇異");.
        }  // EN: Structure delimiter for a block or scope.

        double[] x = new double[n];  // EN: Execute a statement: double[] x = new double[n];.
        for (int j = 0; j < n; j++)  // EN: Loop control flow: for (int j = 0; j < n; j++).
        {  // EN: Structure delimiter for a block or scope.
            double[,] Aj = ReplaceColumn(A, b, j);  // EN: Execute a statement: double[,] Aj = ReplaceColumn(A, b, j);.
            x[j] = Determinant(Aj) / detA;  // EN: Execute a statement: x[j] = Determinant(Aj) / detA;.
        }  // EN: Structure delimiter for a block or scope.
        return x;  // EN: Return from the current function: return x;.
    }  // EN: Structure delimiter for a block or scope.

    static void Main(string[] args)  // EN: Execute line: static void Main(string[] args).
    {  // EN: Structure delimiter for a block or scope.
        PrintSeparator("克萊姆法則示範 (C#)");  // EN: Execute a statement: PrintSeparator("克萊姆法則示範 (C#)");.

        // ========================================
        // 1. 2×2 系統
        // ========================================
        PrintSeparator("1. 2×2 系統");  // EN: Execute a statement: PrintSeparator("1. 2×2 系統");.

        double[,] A2 = {{2, 3}, {4, 5}};  // EN: Execute a statement: double[,] A2 = {{2, 3}, {4, 5}};.
        double[] b2 = {8, 14};  // EN: Execute a statement: double[] b2 = {8, 14};.

        Console.WriteLine("方程組：");  // EN: Execute a statement: Console.WriteLine("方程組：");.
        Console.WriteLine("  2x + 3y = 8");  // EN: Execute a statement: Console.WriteLine(" 2x + 3y = 8");.
        Console.WriteLine("  4x + 5y = 14");  // EN: Execute a statement: Console.WriteLine(" 4x + 5y = 14");.

        PrintMatrix("\nA", A2);  // EN: Execute a statement: PrintMatrix("\nA", A2);.
        PrintVector("b", b2);  // EN: Execute a statement: PrintVector("b", b2);.

        double detA2 = Determinant(A2);  // EN: Execute a statement: double detA2 = Determinant(A2);.
        Console.WriteLine($"\ndet(A) = {detA2:F4}");  // EN: Execute a statement: Console.WriteLine($"\ndet(A) = {detA2:F4}");.

        double[] x2 = CramersRuleSolve(A2, b2);  // EN: Execute a statement: double[] x2 = CramersRuleSolve(A2, b2);.

        for (int j = 0; j < 2; j++)  // EN: Loop control flow: for (int j = 0; j < 2; j++).
        {  // EN: Structure delimiter for a block or scope.
            double[,] Aj = ReplaceColumn(A2, b2, j);  // EN: Execute a statement: double[,] Aj = ReplaceColumn(A2, b2, j);.
            double detAj = Determinant(Aj);  // EN: Execute a statement: double detAj = Determinant(Aj);.
            Console.WriteLine($"\nA{j+1}（第 {j+1} 行換成 b）：");  // EN: Execute a statement: Console.WriteLine($"\nA{j+1}（第 {j+1} 行換成 b）：");.
            PrintMatrix("", Aj);  // EN: Execute a statement: PrintMatrix("", Aj);.
            Console.WriteLine($"det(A{j+1}) = {detAj:F4}");  // EN: Execute a statement: Console.WriteLine($"det(A{j+1}) = {detAj:F4}");.
            Console.WriteLine($"x{j+1} = {x2[j]:F4}");  // EN: Execute a statement: Console.WriteLine($"x{j+1} = {x2[j]:F4}");.
        }  // EN: Structure delimiter for a block or scope.

        Console.WriteLine($"\n解：x = {x2[0]:F4}, y = {x2[1]:F4}");  // EN: Execute a statement: Console.WriteLine($"\n解：x = {x2[0]:F4}, y = {x2[1]:F4}");.

        // ========================================
        // 2. 3×3 系統
        // ========================================
        PrintSeparator("2. 3×3 系統");  // EN: Execute a statement: PrintSeparator("2. 3×3 系統");.

        double[,] A3 = {  // EN: Execute line: double[,] A3 = {.
            {2, 1, -1},  // EN: Execute line: {2, 1, -1},.
            {-3, -1, 2},  // EN: Execute line: {-3, -1, 2},.
            {-2, 1, 2}  // EN: Execute line: {-2, 1, 2}.
        };  // EN: Structure delimiter for a block or scope.
        double[] b3 = {8, -11, -3};  // EN: Execute a statement: double[] b3 = {8, -11, -3};.

        Console.WriteLine("方程組：");  // EN: Execute a statement: Console.WriteLine("方程組：");.
        Console.WriteLine("   2x +  y -  z =  8");  // EN: Execute a statement: Console.WriteLine(" 2x + y - z = 8");.
        Console.WriteLine("  -3x -  y + 2z = -11");  // EN: Execute a statement: Console.WriteLine(" -3x - y + 2z = -11");.
        Console.WriteLine("  -2x +  y + 2z = -3");  // EN: Execute a statement: Console.WriteLine(" -2x + y + 2z = -3");.

        PrintMatrix("\nA", A3);  // EN: Execute a statement: PrintMatrix("\nA", A3);.
        PrintVector("b", b3);  // EN: Execute a statement: PrintVector("b", b3);.

        double[] x3 = CramersRuleSolve(A3, b3);  // EN: Execute a statement: double[] x3 = CramersRuleSolve(A3, b3);.

        Console.WriteLine($"\n解：x = {x3[0]:F4}, y = {x3[1]:F4}, z = {x3[2]:F4}");  // EN: Execute a statement: Console.WriteLine($"\n解：x = {x3[0]:F4}, y = {x3[1]:F4}, z = {x3[2]:F4}"….

        // 驗證
        Console.WriteLine("\n驗證：");  // EN: Execute a statement: Console.WriteLine("\n驗證：");.
        Console.WriteLine($"  2({x3[0]:F0}) + ({x3[1]:F0}) - ({x3[2]:F0}) = {2*x3[0] + x3[1] - x3[2]:F4}");  // EN: Execute a statement: Console.WriteLine($" 2({x3[0]:F0}) + ({x3[1]:F0}) - ({x3[2]:F0}) = {2*x….
        Console.WriteLine($"  -3({x3[0]:F0}) - ({x3[1]:F0}) + 2({x3[2]:F0}) = {-3*x3[0] - x3[1] + 2*x3[2]:F4}");  // EN: Execute a statement: Console.WriteLine($" -3({x3[0]:F0}) - ({x3[1]:F0}) + 2({x3[2]:F0}) = {-….
        Console.WriteLine($"  -2({x3[0]:F0}) + ({x3[1]:F0}) + 2({x3[2]:F0}) = {-2*x3[0] + x3[1] + 2*x3[2]:F4}");  // EN: Execute a statement: Console.WriteLine($" -2({x3[0]:F0}) + ({x3[1]:F0}) + 2({x3[2]:F0}) = {-….

        // 總結
        PrintSeparator("總結");  // EN: Execute a statement: PrintSeparator("總結");.
        Console.WriteLine(@"  // EN: Execute line: Console.WriteLine(@".
克萊姆法則：  // EN: Execute line: 克萊姆法則：.
  xⱼ = det(Aⱼ) / det(A)  // EN: Execute line: xⱼ = det(Aⱼ) / det(A).
  Aⱼ = A 的第 j 行換成 b  // EN: Execute line: Aⱼ = A 的第 j 行換成 b.

適用條件：  // EN: Execute line: 適用條件：.
  - det(A) ≠ 0  // EN: Execute line: - det(A) ≠ 0.
  - 方陣系統  // EN: Execute line: - 方陣系統.

時間複雜度：O(n! × n)  // EN: Execute line: 時間複雜度：O(n! × n).
");  // EN: Execute a statement: ");.

        Console.WriteLine(new string('=', 60));  // EN: Execute a statement: Console.WriteLine(new string('=', 60));.
        Console.WriteLine("示範完成！");  // EN: Execute a statement: Console.WriteLine("示範完成！");.
        Console.WriteLine(new string('=', 60));  // EN: Execute a statement: Console.WriteLine(new string('=', 60));.
    }  // EN: Structure delimiter for a block or scope.
}  // EN: Structure delimiter for a block or scope.

File: 05-determinants/03-cramers-rule/java/CramersRule.java
/**
 * 克萊姆法則 (Cramer's Rule)
 *
 * 編譯：javac CramersRule.java
 * 執行：java CramersRule
 */

public class CramersRule {  // EN: Execute line: public class CramersRule {.

    static void printSeparator(String title) {  // EN: Execute line: static void printSeparator(String title) {.
        System.out.println();  // EN: Execute a statement: System.out.println();.
        System.out.println("=".repeat(60));  // EN: Execute a statement: System.out.println("=".repeat(60));.
        System.out.println(title);  // EN: Execute a statement: System.out.println(title);.
        System.out.println("=".repeat(60));  // EN: Execute a statement: System.out.println("=".repeat(60));.
    }  // EN: Structure delimiter for a block or scope.

    static void printMatrix(String name, double[][] M) {  // EN: Execute line: static void printMatrix(String name, double[][] M) {.
        System.out.println(name + " =");  // EN: Execute a statement: System.out.println(name + " =");.
        for (double[] row : M) {  // EN: Loop control flow: for (double[] row : M) {.
            StringBuilder sb = new StringBuilder("  [");  // EN: Execute a statement: StringBuilder sb = new StringBuilder(" [");.
            for (int i = 0; i < row.length; i++) {  // EN: Loop control flow: for (int i = 0; i < row.length; i++) {.
                sb.append(String.format("%8.4f", row[i]));  // EN: Execute a statement: sb.append(String.format("%8.4f", row[i]));.
                if (i < row.length - 1) sb.append(", ");  // EN: Conditional control flow: if (i < row.length - 1) sb.append(", ");.
            }  // EN: Structure delimiter for a block or scope.
            sb.append("]");  // EN: Execute a statement: sb.append("]");.
            System.out.println(sb.toString());  // EN: Execute a statement: System.out.println(sb.toString());.
        }  // EN: Structure delimiter for a block or scope.
    }  // EN: Structure delimiter for a block or scope.

    static void printVector(String name, double[] v) {  // EN: Execute line: static void printVector(String name, double[] v) {.
        StringBuilder sb = new StringBuilder(name + " = [");  // EN: Execute a statement: StringBuilder sb = new StringBuilder(name + " = [");.
        for (int i = 0; i < v.length; i++) {  // EN: Loop control flow: for (int i = 0; i < v.length; i++) {.
            sb.append(String.format("%.4f", v[i]));  // EN: Execute a statement: sb.append(String.format("%.4f", v[i]));.
            if (i < v.length - 1) sb.append(", ");  // EN: Conditional control flow: if (i < v.length - 1) sb.append(", ");.
        }  // EN: Structure delimiter for a block or scope.
        sb.append("]");  // EN: Execute a statement: sb.append("]");.
        System.out.println(sb.toString());  // EN: Execute a statement: System.out.println(sb.toString());.
    }  // EN: Structure delimiter for a block or scope.

    // 2×2 行列式
    static double det2x2(double[][] A) {  // EN: Execute line: static double det2x2(double[][] A) {.
        return A[0][0] * A[1][1] - A[0][1] * A[1][0];  // EN: Return from the current function: return A[0][0] * A[1][1] - A[0][1] * A[1][0];.
    }  // EN: Structure delimiter for a block or scope.

    // 3×3 行列式
    static double det3x3(double[][] A) {  // EN: Execute line: static double det3x3(double[][] A) {.
        return A[0][0] * (A[1][1] * A[2][2] - A[1][2] * A[2][1])  // EN: Return from the current function: return A[0][0] * (A[1][1] * A[2][2] - A[1][2] * A[2][1]).
             - A[0][1] * (A[1][0] * A[2][2] - A[1][2] * A[2][0])  // EN: Execute line: - A[0][1] * (A[1][0] * A[2][2] - A[1][2] * A[2][0]).
             + A[0][2] * (A[1][0] * A[2][1] - A[1][1] * A[2][0]);  // EN: Execute a statement: + A[0][2] * (A[1][0] * A[2][1] - A[1][1] * A[2][0]);.
    }  // EN: Structure delimiter for a block or scope.

    static double determinant(double[][] A) {  // EN: Execute line: static double determinant(double[][] A) {.
        int n = A.length;  // EN: Execute a statement: int n = A.length;.
        if (n == 2) return det2x2(A);  // EN: Conditional control flow: if (n == 2) return det2x2(A);.
        if (n == 3) return det3x3(A);  // EN: Conditional control flow: if (n == 3) return det3x3(A);.
        throw new RuntimeException("僅支援 2×2 和 3×3 矩陣");  // EN: Execute a statement: throw new RuntimeException("僅支援 2×2 和 3×3 矩陣");.
    }  // EN: Structure delimiter for a block or scope.

    // 替換第 j 行
    static double[][] replaceColumn(double[][] A, double[] b, int j) {  // EN: Execute line: static double[][] replaceColumn(double[][] A, double[] b, int j) {.
        int n = A.length;  // EN: Execute a statement: int n = A.length;.
        double[][] Aj = new double[n][n];  // EN: Execute a statement: double[][] Aj = new double[n][n];.
        for (int i = 0; i < n; i++) {  // EN: Loop control flow: for (int i = 0; i < n; i++) {.
            for (int k = 0; k < n; k++) {  // EN: Loop control flow: for (int k = 0; k < n; k++) {.
                Aj[i][k] = (k == j) ? b[i] : A[i][k];  // EN: Execute a statement: Aj[i][k] = (k == j) ? b[i] : A[i][k];.
            }  // EN: Structure delimiter for a block or scope.
        }  // EN: Structure delimiter for a block or scope.
        return Aj;  // EN: Return from the current function: return Aj;.
    }  // EN: Structure delimiter for a block or scope.

    // 克萊姆法則
    static double[] cramersRule(double[][] A, double[] b) {  // EN: Execute line: static double[] cramersRule(double[][] A, double[] b) {.
        int n = A.length;  // EN: Execute a statement: int n = A.length;.
        double detA = determinant(A);  // EN: Execute a statement: double detA = determinant(A);.

        if (Math.abs(detA) < 1e-10) {  // EN: Conditional control flow: if (Math.abs(detA) < 1e-10) {.
            throw new RuntimeException("矩陣奇異");  // EN: Execute a statement: throw new RuntimeException("矩陣奇異");.
        }  // EN: Structure delimiter for a block or scope.

        double[] x = new double[n];  // EN: Execute a statement: double[] x = new double[n];.
        for (int j = 0; j < n; j++) {  // EN: Loop control flow: for (int j = 0; j < n; j++) {.
            double[][] Aj = replaceColumn(A, b, j);  // EN: Execute a statement: double[][] Aj = replaceColumn(A, b, j);.
            x[j] = determinant(Aj) / detA;  // EN: Execute a statement: x[j] = determinant(Aj) / detA;.
        }  // EN: Structure delimiter for a block or scope.
        return x;  // EN: Return from the current function: return x;.
    }  // EN: Structure delimiter for a block or scope.

    public static void main(String[] args) {  // EN: Execute line: public static void main(String[] args) {.
        printSeparator("克萊姆法則示範 (Java)");  // EN: Execute a statement: printSeparator("克萊姆法則示範 (Java)");.

        // ========================================
        // 1. 2×2 系統
        // ========================================
        printSeparator("1. 2×2 系統");  // EN: Execute a statement: printSeparator("1. 2×2 系統");.

        double[][] A2 = {{2, 3}, {4, 5}};  // EN: Execute a statement: double[][] A2 = {{2, 3}, {4, 5}};.
        double[] b2 = {8, 14};  // EN: Execute a statement: double[] b2 = {8, 14};.

        System.out.println("方程組：");  // EN: Execute a statement: System.out.println("方程組：");.
        System.out.println("  2x + 3y = 8");  // EN: Execute a statement: System.out.println(" 2x + 3y = 8");.
        System.out.println("  4x + 5y = 14");  // EN: Execute a statement: System.out.println(" 4x + 5y = 14");.

        printMatrix("\nA", A2);  // EN: Execute a statement: printMatrix("\nA", A2);.
        printVector("b", b2);  // EN: Execute a statement: printVector("b", b2);.

        double detA2 = determinant(A2);  // EN: Execute a statement: double detA2 = determinant(A2);.
        System.out.printf("%ndet(A) = %.4f%n", detA2);  // EN: Execute a statement: System.out.printf("%ndet(A) = %.4f%n", detA2);.

        double[] x2 = cramersRule(A2, b2);  // EN: Execute a statement: double[] x2 = cramersRule(A2, b2);.

        for (int j = 0; j < 2; j++) {  // EN: Loop control flow: for (int j = 0; j < 2; j++) {.
            double[][] Aj = replaceColumn(A2, b2, j);  // EN: Execute a statement: double[][] Aj = replaceColumn(A2, b2, j);.
            double detAj = determinant(Aj);  // EN: Execute a statement: double detAj = determinant(Aj);.
            System.out.printf("%nA%d（第 %d 行換成 b）：%n", j+1, j+1);  // EN: Execute a statement: System.out.printf("%nA%d（第 %d 行換成 b）：%n", j+1, j+1);.
            printMatrix("", Aj);  // EN: Execute a statement: printMatrix("", Aj);.
            System.out.printf("det(A%d) = %.4f%n", j+1, detAj);  // EN: Execute a statement: System.out.printf("det(A%d) = %.4f%n", j+1, detAj);.
            System.out.printf("x%d = %.4f%n", j+1, x2[j]);  // EN: Execute a statement: System.out.printf("x%d = %.4f%n", j+1, x2[j]);.
        }  // EN: Structure delimiter for a block or scope.

        System.out.printf("%n解：x = %.4f, y = %.4f%n", x2[0], x2[1]);  // EN: Execute a statement: System.out.printf("%n解：x = %.4f, y = %.4f%n", x2[0], x2[1]);.

        // ========================================
        // 2. 3×3 系統
        // ========================================
        printSeparator("2. 3×3 系統");  // EN: Execute a statement: printSeparator("2. 3×3 系統");.

        double[][] A3 = {  // EN: Execute line: double[][] A3 = {.
            {2, 1, -1},  // EN: Execute line: {2, 1, -1},.
            {-3, -1, 2},  // EN: Execute line: {-3, -1, 2},.
            {-2, 1, 2}  // EN: Execute line: {-2, 1, 2}.
        };  // EN: Structure delimiter for a block or scope.
        double[] b3 = {8, -11, -3};  // EN: Execute a statement: double[] b3 = {8, -11, -3};.

        System.out.println("方程組：");  // EN: Execute a statement: System.out.println("方程組：");.
        System.out.println("   2x +  y -  z =  8");  // EN: Execute a statement: System.out.println(" 2x + y - z = 8");.
        System.out.println("  -3x -  y + 2z = -11");  // EN: Execute a statement: System.out.println(" -3x - y + 2z = -11");.
        System.out.println("  -2x +  y + 2z = -3");  // EN: Execute a statement: System.out.println(" -2x + y + 2z = -3");.

        printMatrix("\nA", A3);  // EN: Execute a statement: printMatrix("\nA", A3);.
        printVector("b", b3);  // EN: Execute a statement: printVector("b", b3);.

        double[] x3 = cramersRule(A3, b3);  // EN: Execute a statement: double[] x3 = cramersRule(A3, b3);.

        System.out.printf("%n解：x = %.4f, y = %.4f, z = %.4f%n", x3[0], x3[1], x3[2]);  // EN: Execute a statement: System.out.printf("%n解：x = %.4f, y = %.4f, z = %.4f%n", x3[0], x3[1], x….

        // 驗證
        System.out.println("\n驗證：");  // EN: Execute a statement: System.out.println("\n驗證：");.
        System.out.printf("  2(%.0f) + (%.0f) - (%.0f) = %.4f%n",  // EN: Execute line: System.out.printf(" 2(%.0f) + (%.0f) - (%.0f) = %.4f%n",.
            x3[0], x3[1], x3[2], 2*x3[0] + x3[1] - x3[2]);  // EN: Execute a statement: x3[0], x3[1], x3[2], 2*x3[0] + x3[1] - x3[2]);.
        System.out.printf("  -3(%.0f) - (%.0f) + 2(%.0f) = %.4f%n",  // EN: Execute line: System.out.printf(" -3(%.0f) - (%.0f) + 2(%.0f) = %.4f%n",.
            x3[0], x3[1], x3[2], -3*x3[0] - x3[1] + 2*x3[2]);  // EN: Execute a statement: x3[0], x3[1], x3[2], -3*x3[0] - x3[1] + 2*x3[2]);.
        System.out.printf("  -2(%.0f) + (%.0f) + 2(%.0f) = %.4f%n",  // EN: Execute line: System.out.printf(" -2(%.0f) + (%.0f) + 2(%.0f) = %.4f%n",.
            x3[0], x3[1], x3[2], -2*x3[0] + x3[1] + 2*x3[2]);  // EN: Execute a statement: x3[0], x3[1], x3[2], -2*x3[0] + x3[1] + 2*x3[2]);.

        // 總結
        printSeparator("總結");  // EN: Execute a statement: printSeparator("總結");.
        System.out.println("""  // EN: Execute line: System.out.println(""".

克萊姆法則：  // EN: Execute line: 克萊姆法則：.
  xⱼ = det(Aⱼ) / det(A)  // EN: Execute line: xⱼ = det(Aⱼ) / det(A).
  Aⱼ = A 的第 j 行換成 b  // EN: Execute line: Aⱼ = A 的第 j 行換成 b.

適用條件：  // EN: Execute line: 適用條件：.
  - det(A) ≠ 0  // EN: Execute line: - det(A) ≠ 0.
  - 方陣系統  // EN: Execute line: - 方陣系統.

時間複雜度：O(n! × n)  // EN: Execute line: 時間複雜度：O(n! × n).
            """);  // EN: Execute a statement: """);.

        System.out.println("=".repeat(60));  // EN: Execute a statement: System.out.println("=".repeat(60));.
        System.out.println("示範完成！");  // EN: Execute a statement: System.out.println("示範完成！");.
        System.out.println("=".repeat(60));  // EN: Execute a statement: System.out.println("=".repeat(60));.
    }  // EN: Structure delimiter for a block or scope.
}  // EN: Structure delimiter for a block or scope.
